File: app/src/main/java/com/techyourchance/multithreading/exercises/exercise6/ComputeFactorialUseCase.java
package com.techyourchance.multithreading.exercises.exercise6;

import android.os.Handler;
import android.os.Looper;
import androidx.annotation.WorkerThread;
import com.techyourchance.multithreading.DefaultConfiguration;
import com.techyourchance.multithreading.common.BaseObservable;
import java.math.BigInteger;
import java.util.concurrent.atomic.AtomicReferenceArray;

public class ComputeFactorialUseCase extends BaseObservable<ComputeFactorialUseCase.Listener> {

    private static int MAX_TIMEOUT_MS = DefaultConfiguration.DEFAULT_FACTORIAL_TIMEOUT_MS;
    private final Object LOCK = new Object();
    private final Handler mUiHandler = new Handler(Looper.getMainLooper());
    private int mNumberOfThreads; // safe
    private ComputationRange[] mThreadsComputationRanges; // safe
    private AtomicReferenceArray<BigInteger> mThreadsComputationResults; // safe
    private int mNumOfFinishedThreads = 0; // safe
    private long mComputationTimeoutTime; // safe
    private volatile boolean mAbortComputation; // safe

    void computeFactorial(final int factorialArgument, final String timeoutUserInput) {
        final int timeout = getTimeout(timeoutUserInput);
        new Thread(() -> {
            initComputationParams(factorialArgument, timeout);
            startComputation();
            waitForThreadsResultsOrTimeoutOrAbort();
            processComputationResults();
        }).start();
    }

    private int getTimeout(String userInput) {
        int timeout;
        if (userInput.isEmpty()) {
            timeout = MAX_TIMEOUT_MS;
        } else {
            timeout = Integer.parseInt(userInput);
            if (timeout > MAX_TIMEOUT_MS) {
                timeout = MAX_TIMEOUT_MS;
            }
        }
        return timeout;
    }

    private void initComputationParams(int factorialArgument, int timeout) {
        mNumberOfThreads = factorialArgument < 20
            ? 1 : Runtime.getRuntime().availableProcessors();

        synchronized (LOCK) {
            mNumOfFinishedThreads = 0;
        }

        mAbortComputation = false;

        mThreadsComputationResults = new AtomicReferenceArray<>(mNumberOfThreads);

        mThreadsComputationRanges = new ComputationRange[mNumberOfThreads];

        initThreadsComputationRanges(factorialArgument);

        mComputationTimeoutTime = System.currentTimeMillis() + timeout;
    }

    private void initThreadsComputationRanges(int factorialArgument) {
        int computationRangeSize = factorialArgument / mNumberOfThreads;

        long nextComputationRangeEnd = factorialArgument;
        for (int i = mNumberOfThreads - 1; i >= 0; i--) {
            mThreadsComputationRanges[i] = new ComputationRange(
                nextComputationRangeEnd - computationRangeSize + 1,
                nextComputationRangeEnd
            );
            nextComputationRangeEnd = mThreadsComputationRanges[i].start - 1;
        }

        // add potentially "remaining" values to first thread's range
        mThreadsComputationRanges[0].start = 1;
    }

    @WorkerThread
    private void startComputation() {
        for (int i = 0; i < mNumberOfThreads; i++) {

            final int threadIndex = i;

            new Thread(() -> {
                long rangeStart = mThreadsComputationRanges[threadIndex].start;
                long rangeEnd = mThreadsComputationRanges[threadIndex].end;
                BigInteger product = new BigInteger("1");
                for (long num = rangeStart; num <= rangeEnd; num++) {
                    if (isTimedOut()) {
                        break;
                    }
                    product = product.multiply(new BigInteger(String.valueOf(num)));
                }
                mThreadsComputationResults.set(threadIndex, product);

                synchronized (LOCK) {
                    mNumOfFinishedThreads++;
                    LOCK.notifyAll();
                }
            }).start();
        }
    }

    @WorkerThread
    private void waitForThreadsResultsOrTimeoutOrAbort() {
        synchronized (LOCK) {
            while (mNumOfFinishedThreads != mNumberOfThreads
                && !mAbortComputation
                && !isTimedOut()) {
                try {
                    LOCK.wait(getRemainingMillisToTimeout());
                } catch (InterruptedException e) {
                    return;
                }
            }
        }
    }

    private long getRemainingMillisToTimeout() {
        return mComputationTimeoutTime - System.currentTimeMillis();
    }

    @WorkerThread
    private void processComputationResults() {
        String resultString;

        if (mAbortComputation) {
            resultString = "Computation aborted";
        } else {
            resultString = computeFinalResult().toString();
        }

        // need to check for timeout after computation of the final result
        if (isTimedOut()) {
            resultString = "Computation timed out";
        }

        final String finalResultString = resultString;

        mUiHandler.post(() -> {
            for (Listener listener : getListeners()
            ) {
                listener.onFactorialComputed(finalResultString);
            }
        });
    }

    @WorkerThread
    private BigInteger computeFinalResult() {
        BigInteger result = new BigInteger("1");
        for (int i = 0; i < mNumberOfThreads; i++) {
            if (isTimedOut()) {
                break;
            }
            result = result.multiply(mThreadsComputationResults.get(i));
        }
        return result;
    }

    private boolean isTimedOut() {
        return System.currentTimeMillis() >= mComputationTimeoutTime;
    }

    @Override protected void onLastListenerUnregistered() {
        super.onLastListenerUnregistered();
        mAbortComputation = true;
    }

    interface Listener {
        void onFactorialComputed(String result);
    }
}

File: app/src/main/java/com/techyourchance/multithreading/exercises/exercise10/ComputeFactorialUseCase.kt
package com.techyourchance.multithreading.exercises.exercise10

import androidx.annotation.WorkerThread
import java.math.BigInteger
import java.util.concurrent.TimeUnit
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

class ComputeFactorialUseCase {

    private val reentrantLock = ReentrantLock()
    private val lockCondition = reentrantLock.newCondition()

    private var numberOfThreads: Int = 0
    private var threadsComputationRanges: Array<ComputationRange?> = arrayOf()
    @Volatile private var threadsComputationResults: Array<BigInteger?> = arrayOf()
    private var numOfFinishedThreads = 0

    private var computationTimeoutTime: Long = 0

    private var abortComputation: Boolean = false

    suspend fun computeFactorial(
        argument: Int,
        timeout: Int
    ): Result {

        initComputationParams(argument, timeout)
        startComputation()
        waitForThreadsResultsOrTimeoutOrAbort()
        return processComputationResults()

    }

    private fun initComputationParams(
        factorialArgument: Int,
        timeout: Int
    ) {
        numberOfThreads = if (factorialArgument < 20)
            1
        else
            Runtime.getRuntime().availableProcessors()

        synchronized(reentrantLock) {
            numOfFinishedThreads = 0
            abortComputation = false
        }

        threadsComputationResults = arrayOfNulls(numberOfThreads)

        threadsComputationRanges = arrayOfNulls(numberOfThreads)

        initThreadsComputationRanges(factorialArgument)

        computationTimeoutTime = System.currentTimeMillis() + timeout
    }

    private fun initThreadsComputationRanges(factorialArgument: Int) {
        val computationRangeSize = factorialArgument / numberOfThreads

        var nextComputationRangeEnd = factorialArgument.toLong()
        for (i in numberOfThreads - 1 downTo 0) {
            threadsComputationRanges[i] = ComputationRange(
                nextComputationRangeEnd - computationRangeSize + 1,
                nextComputationRangeEnd
            )
            nextComputationRangeEnd = threadsComputationRanges[i]!!.start - 1
        }

        // add potentially "remaining" values to first thread's range
        threadsComputationRanges[0] = ComputationRange(1, threadsComputationRanges[0]!!.end)
    }

    @WorkerThread
    private fun startComputation() {
        for (i in 0 until numberOfThreads) {

            Thread {
                val rangeStart = threadsComputationRanges[i]!!.start
                val rangeEnd = threadsComputationRanges[i]!!.end
                var product = BigInteger("1")
                for (num in rangeStart..rangeEnd) {
                    if (isTimedOut()) {
                        break
                    }
                    product = product.multiply(BigInteger(num.toString()))
                }
                threadsComputationResults[i] = product

                reentrantLock.withLock {
                    numOfFinishedThreads++
                    lockCondition.signalAll()
                }

            }.start()
        }
    }

    @WorkerThread
    private fun waitForThreadsResultsOrTimeoutOrAbort() {
        reentrantLock.withLock {
            while (numOfFinishedThreads != numberOfThreads && !abortComputation && !isTimedOut()) {
                try {
                    lockCondition.await(remainingMillisToTimeout(), TimeUnit.MILLISECONDS)
                } catch (e: InterruptedException) {
                    return
                }

            }
        }
    }

    @WorkerThread
    private fun processComputationResults(): Result {
        if (abortComputation) {
            return Result.Aborted
        }

        val result = computeFinalResult()

        // need to check for timeout after computation of the final result
        if (isTimedOut()) {
            return Result.Timeout
        }

        return Result.Factorial(result)
    }

    @WorkerThread
    private fun computeFinalResult(): BigInteger {
        var result = BigInteger("1")
        for (i in 0 until numberOfThreads) {
            if (isTimedOut()) {
                break
            }
            result = result.multiply(threadsComputationResults[i])
        }
        return result
    }

    private fun remainingMillisToTimeout(): Long {
        return computationTimeoutTime - System.currentTimeMillis()
    }

    private fun isTimedOut(): Boolean {
        return System.currentTimeMillis() >= computationTimeoutTime
    }

    private data class ComputationRange(
        val start: Long,
        val end: Long
    )

    sealed class Result {
        object Timeout : Result() {
            override fun toString(): String {
                return "Computation timed out"
            }
        }

        object Aborted : Result() {
            override fun toString(): String {
                return "Computation was aborted"
            }
        }

        class Factorial(private val bigInteger: BigInteger) : Result() {
            override fun toString(): String {
                return bigInteger.toString()
            }
        }
    }
}
